File: 1-Relational Database/1-Relational DB Concepts/B-Tree & B+ Tree/java/BTree.java
import common_constructs.IndexEntry;
import common_constructs.Node;

public class BTree extends BTreeBase {

    /**
     * Constructor with parameter.
     * @param order order of the B-tree
     */
    public BTree(int order) {
        super(order);
    }

    protected String searchHelper(Node curr, int key) {
        int pos = curr.findInsertPos(key);
        if (pos >= 0) { // Found it
            return ((IndexEntry) curr.getEntry(pos)).getRecordAddress();
        }
        if (curr.isLeaf()) { // Leaf
            return null;
        }
        // Non-leaf
        // Go to the appropriate child
        int insertPos = -pos - 1;
        return searchHelper(curr.getChild(insertPos), key);
    }

}

File: 1-Relational Database/1-Relational DB Concepts/B-Tree & B+ Tree/python/btree.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
A very simple B-tree implementation.
"""

__author__ = 'Ziang Lu'

from typing import Optional

from btree_base import BTreeBase
from common_constructs import Node


class BTree(BTreeBase):
    __slots__ = []

    def __init__(self, order: int):
        """
        Constructor with parameter.
        :param order: int
        """
        super().__init__(order)

    def _search_helper(self, curr: Node, key: int) -> Optional[str]:
        pos = curr.find_insert_pos(key)
        if pos < curr.size() and key == curr.get_entry(pos).key:  # Found it
            return curr.get_entry(pos).record_address
        if curr.is_leaf():  # Not found
            return None
        # Non-leaf
        # Go to the appropriate child
        return self._search_helper(curr.get_child(pos), key)
